### PR TITLE
nixos/geth: initial service

### PR DESCRIPTION
**Pull request by bachp**
###### Motivation for this change

Add a service to run a Ethereum node with go-ethereum.
The service uses the same pattern as the [bitcoind](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/bitcoind.nix) and allows multiple instances to be run.

Contributes to https://github.com/NixOS/nixpkgs/issues/103890

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+  gethOpts = { config, lib, name, ...}: {
+
+    options = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable geth node.";
+      };
```
Comment (bachp) [inline nixos/modules/services/blockchain/ethereum/geth.nix:16]: Just noticed I should have used `lib.mkEnableOption` here.

**Comment by excogitation** [comment]
Did a quick test synching goerli. Looks to be working.
Screenshot from 2021-01-03 01-45-46

Wouldn't it make sense to have "ipcdisable" as an option to be able to enable attaching to geth?

**Comment by bachp** [comment]
@excogitation I ussually attacht to the node over http via `geth attach http://localhost:8545`. Adding an option would sure be possible. Do you see an advantage in having IPC too? The option could also be added later should the need arise.

As the module supports multiple instances, it might also be worth to create a wrapper script `geht-<instance>`, that has the correct parameters already set.

**Comment by bachp** [comment]
@RaghavSood @Mic92 Could you take a look at this MR?

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      extraArgs = mkOption {
+        type = types.str;
+        description = ''
+          Additional arguments passed to Go Ethereum.
+        '';
+        default = "";
+      };
+
+      package = mkOption {
+        default = pkgs.go-ethereum;
```
Comment (attila-lendvai) [inline nixos/modules/services/blockchain/ethereum/geth.nix:130]: the package is called `go-ethereum`. shouldn't we call the service the same?

**Review by attila-lendvai — CHANGES_REQUESTED** [review]

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 2 above)
Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix:130]: Looking at [other distros](https://repology.org/project/go:ethereum/versions) there seems to be no consesus if the package should be called geth or go-ethereum.

However I think we should name the service `geth` as that is the name you find in almost all guides including a popular [eth2 staking guide](https://someresat.medium.com/guide-to-staking-on-ethereum-2-0-ubuntu-lighthouse-41de20513b12). I think this is what most people expect.

I would rather propose that we rename the package to `geth` to make it more consistent.

Comment (attila-lendvai) [reply nixos/modules/services/blockchain/ethereum/geth.nix:130]: fair enough. i don't have a strong opinion on which of the two should it be, but i think being the same is preferable.

Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix]: The `go-ethereum` package is now renamed to `geth`.

Pinging maintainers for approval @adisbladis @lionello @xrelkd @RaghavSood

**Comment by adisbladis** [comment]
I do not want to rename the package to to `geth`.
Geth is just one of the binaries in this package and the upstream repository name is `go-ethereum`.

I would be fine with an alias mapping `geth` to `go-ethereum`.

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+          type = types.port;
+          default = 6060;
+          description = ''
+            Port number of Go Ethereum metrics service.
+          '';
+        };
+      };
+
+      network = mkOption {
+        type = types.enum [ "mainnet" "goerli" "inkeby" "yolov2" "ropsten" ];
```
Comment (adisbladis) [inline nixos/modules/services/blockchain/ethereum/geth.nix:98]:
```suggestion
        type = types.enum [ "mainnet" "goerli" "rinkeby" "yolov2" "ropsten" ];
```

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      gcmode = mkOption {
+        type = types.enum [ "full" "archive" ];
+        default = "full";
+        description = ''
+           Blockchain garbage collection mode.
+        '';
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
```
Comment (adisbladis) [inline nixos/modules/services/blockchain/ethereum/geth.nix:122]:
```suggestion
        type = types.listOf types.str;
        description = "Additional arguments passed to Go Ethereum."
        default = [ ];
```

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        NoNewPrivileges = "true";
+        PrivateDevices = "true";
+        MemoryDenyWriteExecute = "true";
+      };
+
+      script = ''
+        ${cfg.package}/bin/geth \
+          --nousb \
+          --ipcdisable \
+          ${optionalString (cfg.network == "goerli") "--goerli"} \
```
Comment (adisbladis) [inline nixos/modules/services/blockchain/ethereum/geth.nix:179]: I'd make this one bif if statement which errors out on unknown networks, that will prevent any future programming mistakes in this module.

**Review by adisbladis — CHANGES_REQUESTED** [review]

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 5 above)
Comment (adisbladis) [reply nixos/modules/services/blockchain/ethereum/geth.nix:179]: Or maybe just simply `--${cfg.network}` as that seems OK for all currently defined networks.

Comment (adisbladis) [reply nixos/modules/services/blockchain/ethereum/geth.nix:179]: Or... Scratch that, this module should also be able to handle `cfg.network` being `null` which might be a reasonable thing for private networks.

Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix:179]: @adisbladis My proposal would be to change it to either: 
1. `types.nullOr types.string`
2. `types.nullOr types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ]`

Do you have a preference?

1. is more flexible as it allows any protential network without extending the service. 2. Is safer but requires to update the service in lockstep with the package.

Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix]: Per @adisbladis comment it makes sense to keep both service and package as is. As the geth service is just one think provided by the go-ethereum package.

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 2 above)
Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix:130]: Per @adisbladis comment it makes sense to keep both service and package as is. As the geth service is just one think provided by the go-ethereum package.

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 5 above)
Comment (adisbladis) [reply nixos/modules/services/blockchain/ethereum/geth.nix]: I think I prefer option 2. If you want to use a network which is not defined in the module you can always use `extraOptions`.

**Comment by adisbladis** [comment]
@bachp I decided to make `go-ethereum` multi output in https://github.com/NixOS/nixpkgs/pull/113960 so this service doesn't need to pull in the entirety of `go-ethereum` as a runtime dependency.

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 5 above)
Comment (bachp) [reply nixos/modules/services/blockchain/ethereum/geth.nix]: It's now `types.nullOr (types.enum [ "goerli" "rinkeby" "yolov2" "ropsten" ])`

**Comment by bachp** [comment]
I did the following changes:
- package now defaults to `go-ethereum.geth`
- package is now added to the systemPackages to be able to interact with the daemon
- added test that checks if the daemon starts up with the correct network

**Comment by adisbladis** [comment]
All good!

**Pull request merged**